core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/flow/AmbariClusterConnector.java

```diff
@@ -491,18 +493,56 @@ private PollingResult restartHadoopServices(Stack stack, AmbariClient ambariClie
         }
         if (serviceComponents.containsKey("NAGIOS")) {
-            restartRequests.put("NAGIOS", ambariClient.restartServiceComponents("NAGIOS", Arrays.asList("NAGIOS_SERVER")));
+            restartRequests.put("NAGIOS", restartComponentWithClient(ambariClient, "NAGIOS", "NAGIOS_SERVER"));
         }
         if (serviceComponents.containsKey("GANGLIA")) {
-            restartRequests.put("GANGLIA", ambariClient.restartServiceComponents("GANGLIA", Arrays.asList("GANGLIA_SERVER")));
+            restartRequests.put("GANGLIA", restartComponentWithClient(ambariClient, "GANGLIA", "GANGLIA_SERVER"));
         }
         return waitForAmbariOperations(stack, ambariClient, restartRequests);
     }
 
+    private int restartComponentWithClient(AmbariClient ambariClient, String service, String... compontents) {
+        return ambariClient.restartServiceComponents(service, Arrays.asList(compontents));
+    }
+
+    private void startServiceIfNeeded(Stack stack, AmbariClient ambariClient, String blueprint) throws CloudbreakException {
+        Map<String, String> componentsCategory = ambariClient.getComponentsCategory(blueprint);
+        Map<String, Map<String, String>> hostComponentsStates = ambariClient.getHostComponentsStates();
+        for (Map.Entry<String, Map<String, String>> hostComponentsEntry : hostComponentsStates.entrySet()) {
+            Map<String, String> componentStateMap = hostComponentsEntry.getValue();
+            List<String> services = new ArrayList<>();
+            for (Map.Entry<String, String> componentStateEntry : componentStateMap.entrySet()) {
+                String category = componentsCategory.get(componentStateEntry.getKey());
+                if (!"CLIENT".equals(category)) {
+                    if ("INSTALLED".equals(componentStateEntry.getValue())) {
+                        if ("NODEMANAGER".equals(componentStateEntry.getKey())) {
+                            services.add("NODEMANAGER");
+                        } else if ("DATANODE".equals(componentStateEntry.getKey())) {
+                            services.add("DATANODE");
+                        } else if ("HBASE_REGIONSERVER".equals(componentStateEntry.getKey())) {
+                            services.add("HBASE_REGIONSERVER");
+                        } else {
+                            LOGGER.warn("No need to restart ambari service: {}", componentStateEntry.getKey());
+                        }
+                    } else {
+                        LOGGER.warn("Ambari service already running: {}", componentStateEntry.getKey());
+                    }
+                }
+            }
+            if (!services.isEmpty()) {
+                try {
+                    Map<String, Integer> stringIntegerMap = ambariClient.startComponentsOnHost(hostComponentsEntry.getKey(), services);
+                    waitForAmbariOperations(stack, ambariClient, stringIntegerMap);
```

Comment:
With this behavior we send the start request and wait, send another request and wait. I think we should collect the requestIds outside the loop and only wait for them once we sent all the requests, just like in case of decommission.